src/main/java/com/google/devtools/build/lib/bazel/rules/python/BazelPythonSemantics.java

```diff
@@ -370,9 +370,10 @@ private static void createPythonZipAction(
     for (Artifact artifact : runfilesSupport.getRunfilesArtifacts().toList()) {
       if (!artifact.equals(executable) && !artifact.equals(zipFile)) {
         argv.addDynamicString(
-            getZipRunfilesPath(artifact.getRunfilesPath(), workspaceName, legacyExternalRunfiles)
-                + "="
-                + artifact.getExecPathString());
+            (getZipRunfilesPath(artifact.getRunfilesPath(), workspaceName, legacyExternalRunfiles)
+                    + "="
+                    + artifact.getExecPathString())
+                .intern());
```

Comment:
You should use BlazeInterners.weakInterner for this, like the rest of the codebase. https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/concurrent/BlazeInterners.java;l=38;drc=ec8fb7969614c6b3752392089e2310708d103225

Interning is not available in Starlark, so this will have to be optimised in a different way.

Even better would be if you replaced call `argv.addDynamicString` with one of the calls that accepts artifact.getExecPath() as an argument and set format to `getZipRunfilesPath(artifact.getRunfilesPath(), workspaceName, legacyExternalRunfiles) + "="`.  

Can you also provide some benchmarking results? baseline vs the change?
I'm mostly interested in used-heap-size-after-gc displayed by `bazel info` after building the project. (You should do 1 warm-up run before collecting the results).

Unfortunately I don't have appropriate project to run benchmarks for you.

Reply:
Nice results!

Those can be better + Starlarkifiable. Using lazy argument construction, Bazel wouldn't retain any of those strings.

Change the whole for loop to something like:
`argv.addAll(VectorArg.of(runfilesSupport.getRunfilesArtifacts()).mapped(MAP_FN))`

where
```
MAPFN = (artifact, args) ->
{ 
if (!artifact.equals(executable) && !artifact.equals(zipFile)) {
          args.accept(getZipRunfilesPath(artifact.getRunfilesPath(), workspaceName, legacyExternalRunfiles)
                    + "="
                    + artifact.getExecPathString());
     }
}
```

Example:
https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/rules/java/OneVersionCheckActionBuilder.java;l=121-128;drc=103ea9d3f3a0d85a0d7c1ad98fd8dbb422505d95

Reply:
Implemented second solution by adding `CommandLineItem.DefinedParamsCacheMapFn` , PTAL

Reply:
I removed `CapturingMapFn`, because it was unused. I think the solution using it would be nice. Let me rollback this removal.

Reply:
A sec, before I rollback removal of `CapturingMapFn`.

> Using ParametrizedMapFn requires specifying INT_MAX in maxInstancesAllowed

When implementing this in Starlark it uses ParametrizedMapFn with maxInstancesAllowed set to Integer.MAX_VALUE. See https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/analysis/starlark/StarlarkCustomCommandLine.java;l=893-976;drc=8f0034ce1e7854521e259a5eaf71859b1e6f95bc

The requirement in ParametrizedMapFn is:
```
  * <p>The user promises that the number of distinct instances constructed is closer to O(rule
   * class count) than O(rule count).
```

In any case, I'd prefer if we don't add new things to `CommandLineItem` and `NestedSetFingerprintCache`

Reply:
Rolled back removal of `CapturingMapFn` in https://github.com/bazelbuild/bazel/commit/c70e0721f0ed506fad62ab89561804a60ab9b976

Reply:
> @comius friendly ping. Do you think we can proceed with this PR?

There's a second PR that handles this with `CapturingMapFn`: https://github.com/bazelbuild/bazel/pull/15037